### PR TITLE
Integrate Inductor with oneDNN Graph fusions (currently only for CPU)

### PR DESCRIPTION
**Pull request by sanchitintel**
####  Platforms supported
only builds on Linux with `gcc version <= 10` but once oneDNN version would be updated to oneDNN v3.5 (currently, v3.4.2 is being used), gcc 11 can also be used. PyTorch is released with gcc 9, and the binary can also be used on systems with gcc 11 toolchain. The gcc version limitation is only for builds (with binaries built with gcc11, a symbol resolution issue would occur during dynamic linking, but that issue has been fixed in oneDNN v3.5) . 
`AVX512_VNNI`, `AVX512_BF16`, `AMX_INT8` & `AMX_BF16` CPU ISAs are supported.  BF16 (or `int8_bf16`) fusions need at least AVX512_BF16 ISA support.

#### Limitations
Only static shapes are currently supported, so it's recommended to do a warmup with all possible shapes that may be encountered over the lifetime of a workload, before measuring latency/throughput.

#### How to use
Inductor configs `freezing` & `onednn_graph` need to be set to true. By default, this feature is disabled.

#### Integration design
An FX subgraph (some nodes of an FX graph) are replaced by an FX node corresponding to an opaque python callable.
This PR introduces a class `InductorImplicitFallback`, which could be reused in the future by others as well.

#### Patterns
This base PR only accelerates [a subgraph containing BERT MHA](https://github.com/oneapi-src/oneDNN/blob/main/src/graph/backend/graph_compiler/patterns/mha_pattern.hpp#L1611-L1637) with `int8_fp32` & `int8_bf16` quantization.
This particular one can't be handled in `torch/_inductor/fx_passes/fuse_attention.py`, because the implementation creates an opaque python callable (If I'd create it in a replacement function, it'd be re-created each time). The corresponding pattern & code are in `torch/_inductor/fx_passes/quantization.py`.
Other accompanying PRs would accelerate subgraphs with DistilBERT int8 MHA, ViT int8 MHA, [LLaMA BF16 (with AMP) MLP pattern](https://github.com/oneapi-src/oneDNN/blob/main/src/graph/backend/graph_compiler/patterns/mlp_pattern.hpp#L319-L343).

PyTorch Quantization 2.0 doesn't currently support LLMs, so they can't be accelerated at this point.

#### Benchmark results
For this specific PR, various BERT-based models were benchmarked:
image

#### UTs
`test/test_onednn_graph_fusions.py` -> BERT UT for int8_fp32 adapted from https://github.com/pytorch/pytorch/pull/124832. Enabled for int8_bf16 as well. SDPA fusions would need:
```python
quantizer.set_function_type_qconfig(
            torch.matmul, quantizer.get_global_quantization_config()
        )
``` 

cc @jgong5 @mingfeima @XiaobingSuper @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @voznesenskym @penguinwu @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by sanchitintel** [comment]
Hi @jansel @eelison, 

in this PR, I'm trying to add
1. Integration code for oneDNN Graph fusions to work with Inductor (disabled by default)
2. A fusion pattern to verify that (1) works. I chose a fusion corresponding to [a subgraph containing quantized BERT MHA](https://github.com/oneapi-src/oneDNN/blob/main/src/graph/backend/graph_compiler/patterns/mha_pattern.hpp#L1611-L1637) pattern. 

Please advise if I should submit a ghstack with all other patterns as well, because `1` is needed by other patterns I intend to add.

Additionally, at some point, the default CPU backend may also support a similar pattern. However, I can't add a replacement for it in `torch/_inductor/fx_passes/fuse_attention.py`, since I also need to add a new node in the `torch.fx` subgraph (besides the replacement). So, I'm adding that replacement in a freezing pass in `torch/_inductor/fx_passes/quantization.py` instead.

As for serializing the pattern, `torch/_inductor/fx_passes/fuse_attention.py` requires a pattern & a replacement to be defined, but currently, it's not possible to add a performant replacement in the default path. So, please advise if we could merge #124832 (which adds a pattern `2` could use) after using a `check_fn` that'd cause the replacement defined in #123832 to not be used, so it would at least help in serializing that pattern `2` in this PR could use.

Thanks!

**Review by jansel — CHANGES_REQUESTED** [review]
I don't think just calling a different graph compiler is a good way to improve TorchInductor.  It creates bifurcation between CPU/GPU backends which will be hard to maintain.

I'd suggest instead making a oneDNN backend for `aten.scaled_dot_product_attention`, and extending that API to support quantization in a way that will apply to both CPU and GPU.

**Comment by sanchitintel** [comment]
Hi @jansel, thanks for your feedback!

In the future, oneDNN Graph would also support XPU (Intel GPU). However, Graph Compiler wouldn't be exposed to PyTorch, i.e. from PyTorch perspective, it'd be a black box. 

>  It creates bifurcation between CPU/GPU backends which will be hard to maintain.

I understand your concern about bifurcation from a code-management perspective, but even now, CPU & GPU passes differ a lot since the GPU ones also use triton. Even if oneDNN Graph would not be used, future custom fusions for CPU would further deviate from those used for CUDA, anyway.


> I'd suggest instead making a oneDNN backend for aten.scaled_dot_product_attention, and extending that API to support quantization in a way that will apply to both CPU and GPU

Sorry, that's not possible because:

1. oneDNN Graph has ~30 subgraphs corresponding to MHA/SDPA of various models & datatypes used. 
The PyTorch SDPA op is generic SDPA (without adjoining ops), which corresponds to just one of those 30 subgraphs.
Such an approach (separate backend for `aten.scaled_dot_product_attention`) would not allow us to accelerate SDPA for most models with oneDNN Graph.

2. Moreover, in the future, oneDNN Graph would be able to fuse as many ops as possible as part of a max-partition policy.
Again, the use of `aten.scaled_dot_product_attention` would allow only one of tens of oneDNN Graph SDPA patterns to be leveraged!

3. oneDNN Graph doesn't just support SDPA. It also supports CNNs & MLP fusions. For LLaMA, it even has a [custom fusion pattern with RMSNorm & MLP fused](https://gist.githubusercontent.com/sanchitintel/7ad097c825acfc008ebc14cfadebafd5/raw/42bba4a373318749c8765bbdec5a511ce087ffcd/LLaMA_MLP_visualization).

Thanks!

**Comment by sanchitintel** [comment]
There’s only one oneDNN Graph MHA pattern that corresponds to generic MHA (without any transpose & reorders).
But that pattern of `matmul  -> scale (optional) -> attention_mask (optional) -> softmax -> matmul` only exists for FP32 dtype.

For BF16 & int8, a pattern of `matmul + softmax + matmul` also exists (also has dequantize/quantize & typecasts in int8_fp32 and int8_bf16 cases), but this pattern does not correspond to Scaled Dot product Attention.

Hence, at this point, I won’t be able to proceed with integrating oneDNN Graph MHA as a fast-path in ATen.

Closing this PR for now. Thanks!

**Comment by sanchitintel** [comment]
Will reopen when the next version of oneDNN will be integrated with PyTorch. Thanks!
